pytorch_lightning/utilities/auto_restart.py

```diff
@@ -0,0 +1,236 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, Dict, Generator, Iterator, List, Optional, Union
+
+from torch.utils.data import BatchSampler, get_worker_info, Sampler
+from torch.utils.data.dataloader import IterableDataset
+
+from pytorch_lightning.utilities.enums import AutoRestartBatchKeys
+
+
+class FastForwardSampler(Sampler):
+    """
+    This FastForwardSampler wraps a :class:`torch.utils.data.Sampler` and records the number of iterations
+    performed during an epoch. It maintains a state through :meth:`state_dict` that can be reloaded through
+    :meth:`load_state_dict`. If the sampler is used in a multiprocessing context, the FastForwardSampler will record
+    the state of the current worker.
+
+    When reloading, the FastForwardSampler will `fast-forward` the wrapped sampler by iterating through all the
+    samples seen in the last iterations (for the current worker).
+    """
+
+    def __init__(self, sampler: Union[Sampler, BatchSampler, Generator]) -> None:
+        super().__init__(data_source=None)
+        self.sampler = sampler
+        self.restarting: bool = False
+        self._current_iteration = 0
+        self._dataloader_batch_size: Optional[int] = None
+        self._cached_state_dict: Optional[Dict[str, Any]] = None
+
+    def setup(self, dataloader_batch_size: Optional[int] = None) -> None:
+        """
+        Setup the ``FastForwardSampler``.
+        This is required only when the provided dataset subclassed :class:`torch.utils.data.Dataset`.
+        """
+        self._dataloader_batch_size = dataloader_batch_size
+
+    @property
+    def worker_id(self) -> int:
+        worker_info = get_worker_info()
+        return worker_info.id if worker_info else 0
+
+    @property
+    def drop_last(self) -> bool:
+        return self.sampler.drop_last
+
+    @property
+    def batch_size(self) -> int:
+        return self.sampler.batch_size
```

Comment:
What kind of sampler do we expect here? It usually doesn't have a batch size, batch_indices, drop last etc?

We should make it clear in the docs what type of samplers this FFSampler can wrap

Reply:
The FastForwardSampler can wrap both the `BatchSampler` or `Sampler` and therefore need to expose the sampler attributes.
Should we have a smarter `getattr` ? If the key doesn't exist in __dict__, then use self.sampler to get the value. Might be better.

Reply:
I see. I think a simple instance check against BatchSampler would do, and return None otherwise.

Reply:
And which type of sampler has an attribute "batch_indices"?

Reply:
The IndexBatchSampler for `predict`.

Reply:
Removed for getattr